Feature migrate nonprofit timezone names

jrmehle opened the pull request:
**NOTE: DO NOT discuss internal CommitChange information in your PR; this PR will be public.
Link back to the issue in the Tix repo when you need to do that.**

~~Swaps to using the built-in Rails `time_zone_select` for Nonprofit timezone field. We were previously using the IANA name off of tzinfo, but the select uses Rails' internal name.~~

~~Added a maintenance task to migrate nonprofit records from the old naming scheme to the new one. Once I had created the maintenance task, I found that the routes wouldn't allow the route to the individual task show page: `http://localhost:5000/maintenance_tasks/tasks/Maintenance::UpdateNpTimezoneNamesTask` and seemed to collide with the nonprofit route:
`get ":state_code/:city/:name" => "nonprofits#show", :as => :nonprofit_location`~~

Initially was going to change the name stored in the database for timezones from the IANA name to Rails internal name so that `time_zone_select` could be used. However, found that there are some database operations where Postgres directly interacts with the timezone field. Postgres can only work with so many formats and Rails internal format is not one of them unfortunately.

That said, some of the work in here is useful and we can still utilize Rails to generate the select options in a way that can be reused for events.

Consider this pre-work in preparation for events having timezones.

jrmehle committed:
Use Rails time zone names

Match those used in in time zone select helper

Uses a maintenance task to convert from the existing IANA names to Rails names.

jrmehle committed:
Move maintenance tasks up in routes to avoid error

Otherwise /maintenance_tasks is matched by:
get ":state_code/:city/:name" => "nonprofits#show", :as => :nonprofit_location

wwahammy commented:
> Added a maintenance task to migrate nonprofit records from the old naming scheme to the new one. Once I had created the maintenance task, I found that the routes wouldn't allow the route to the individual task show page: `http://localhost:5000/maintenance_tasks/tasks/Maintenance::UpdateNpTimezoneNamesTask` and seemed to collide with the nonprofit route: `get ":state_code/:city/:name" => "nonprofits#show", :as => :nonprofit_location`

This worries me a bit but your fix is probably sufficient. Could you add a route test to verify that the routing we want continues to be the case? We know it works now but my fear is someone could move the route at some point and, without knowing why it happened, we'd struggle to debug it.

wwahammy reviewed (commented):
Outside of what I mentioned before, this seems fine other than the specs needing to be fixed.

One question: Have you run this maintenance task locally on a copy of the live DB? Just want to make sure that's covered.

jrmehle committed:
Add spec around maintenance tasks

Also included a small refactor to move some code common to all routing specs into a more central place.

jrmehle commented:
> > Added a maintenance task to migrate nonprofit records from the old naming scheme to the new one. Once I had created the maintenance task, I found that the routes wouldn't allow the route to the individual task show page: `http://localhost:5000/maintenance_tasks/tasks/Maintenance::UpdateNpTimezoneNamesTask` and seemed to collide with the nonprofit route: `get ":state_code/:city/:name" => "nonprofits#show", :as => :nonprofit_location`
> 
> This worries me a bit but your fix is probably sufficient. Could you add a route test to verify that the routing we want continues to be the case? We know it works now but my fear is someone could move the route at some point and, without knowing why it happened, we'd struggle to debug it.

No problem. I added a spec to cover at least the task index and show routes. Looking at the gem code, there are [some other routes](https://github.com/Shopify/maintenance_tasks/blob/v2.12.0/config/routes.rb) could be tested too, but they're nested under the task so the likelihood we'd run into collisions in route definition is less and so I opted to not do that for now.

jrmehle commented:
> Outside of what I mentioned before, this seems fine other than the specs needing to be fixed.
> 
> One question: Have you run this maintenance task locally on a copy of the live DB? Just want to make sure that's covered.

I have and it worked at first glance, but I need to do a more thorough investigation to see if any of the old names didn't convert.

jrmehle committed:
Remove maintenance task

jrmehle committed:
Switch back but make time zone select helper

Reverted the change to storing Rails timezone names in the database due to some of the custom query code directly working with timezones in postgres.

Opted instead for a helper that outputs the timezones in a similar fashion for the end user in the select even though it is not the Rails helper, it can be used for when events get timezones added.

jrmehle committed:
Remove spec file that isn't covering anything now

But keep directory around.

jrmehle committed:
Merge branch 'supporter_level_goal' into feature/jaredmehle/migrate-nonprofit-timezone-names

jrmehle committed:
Merge branch 'supporter_level_goal' into feature/jaredmehle/migrate-nonprofit-timezone-names

jrmehle committed:
Merge branch 'supporter_level_goal' into feature/jaredmehle/migrate-nonprofit-timezone-names

wwahammy commented on app/helpers/timezone_helper.rb:
```diff
@@ -0,0 +1,9 @@
+module TimezoneHelper
+  def time_zone_options_with_iana(selected = nil)
+    zones = ActiveSupport::TimeZone.all.map do |zone|
+      [zone.to_s, zone.tzinfo.name] # label, IANA value
+    end
+
+    options_for_select(zones, selected)
+  end
```
I feel like this should have a spec but it's pretty simple. Thoughts?

jrmehle replied:
No problem, added a spec.

wwahammy reviewed (commented):
For the most part, it looks good. Just one question.

jrmehle committed:
Spec coverage for the newly added TimezoneHelper

wwahammy reviewed (approved):
Looks good!
